src/averageAges.js

```diff
@@ -54,11 +65,20 @@ function calculateWomenAverageAge(people, withChildren) {
  *
  * @return {number}
  */
+
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const children = people.map((person) => (
+    {
+      ...person,
+      motherInfo: people
+        .find((motherName) => motherName.name === person.mother),
+    }))
+    .filter(person => person.motherInfo && (onlyWithSon === undefined
+  || person.sex === 'm'));
```

Comment:
It's written correctly now, but it's hard to read and maintain in the future.
We can break the code into several parts:
1. Filter the `people` array and leave only those objects in which there are `mothers`.
2. If `onlyWithSon` is `true` - filter the `children` array and leave only `male` objects there.
3. When calculating the `average age`, `find` for the `mother` for each `child` in the middle of the `reduce` method and then perform the calculation.